### PR TITLE
Add S3KeysUnchangedSensor

### PR DESCRIPTION
**Pull request by anikarni**
I had a use for this sensor in project work, so I figured I'd contribute it to the repo. The S3UploadSessionCompleteSensor is essentially the same implementation as exists for GCS. The only things that changes is the hook call.

The S3UploadSessionCompleteSensor polls the number of files in a S3 bucket and verifies against the previous count. If the count does not change for a given interval, the sensor succeeds. 

Since the logic for the sensor is the same as GCS, I considered refactoring for code reuse, but I don't exactly understand how provider package dependencies work, so I opted for duplicating the code. Feel free to suggest an alternative approach if this does not seem adequate.

Thanks!

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py
Comment (ashb) [inline airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]: It should throw an error then.

Comment (anikarni) [reply airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]: It does! The `@poke_mode_only` decorator should do the job!

File: airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py
Comment (ashb) [inline airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]: We're already using https://pypi.org/project/freezegun/ in the tests, so you don't need to alter the code to make it easier to test

**Comment by anikarni** [comment]
hey @ashb , I changed the tests to use `freeze` and also added a test to show that the sensor throw on `reschedule` mode. thanks for giving it a look ;)

File: airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py
Comment (ashb) [inline airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]:
```suggestion
    increase in the number of objects. Note, this sensor will not behave correctly
```

File: airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py
Comment (ashb) [inline airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]:
```suggestion

    @cached_property
    def hook(self):
        return S3Hook(aws_conn_id=self.aws_conn_id, verify=self.verify)
```

(You'll need `from cached_property import cached_property` too)

File: airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py
Comment (ashb) [inline airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]:
```suggestion
                self.log.info(
```

I don't think that feels like a warning -- the user has said deleted files are okay.

File: airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py
Comment (ashb) [inline airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]:
```suggestion
                    "Objects were deleted during the last poke interval. Updating the file counter and resetting "
                    "last_activity_time:\n    %s",
                    self.previous_objects - current_objects
```

File: airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py
Comment (ashb) [inline airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]: This method is called `is_bucket_updated`, and it has been here, but we are returning False -- Given the method name, I would expect this to be True here.

File: airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py
Comment (ashb) [inline airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]: A few lines up we have already done `self.previous_objects = current_objects` -- this is always going to return an empty set :)

File: tests/providers/amazon/aws/sensors/test_s3_upload_session_complete.py
Comment (ashb) [inline tests/providers/amazon/aws/sensors/test_s3_upload_session_complete.py]:
```suggestion
                poke_interval=0.1,
```

We want our tests to run quickly!

**Review by ashb — DISMISSED** [review]
We possibly need to rename this. On looking at again with fresh eyes I thought this might be to do with MultiPartUploads, but it isn't. I can't really think of a "clearer" name though. Perhaps "S3BucketUnchangedSensor"? "S3KeysUnchangedSensor"? "S3InactivtySensor"? Not 100% sold on any of these names.

Additionally we aren't ever testing `poke()` directly -- at least one test using that runs poke (likely by mocking the return value of `sensor.hook.list_keys`) would be good please.

Also please note: I'm going to be away for six weeks at the end of today, so I wont' be able to re-review this. Committers: feel free to dismiss my review once the points have been addressed to your satisfaction.

File: airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py
Comment (ashb) [reply airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]: (You can either change the return value, or change the name)

File: tests/providers/amazon/aws/sensors/test_s3_upload_session_complete.py
Comment (ashb) [reply tests/providers/amazon/aws/sensors/test_s3_upload_session_complete.py]: Oh this possibly doesn't matter as it's only run once?

Comment (anikarni) [reply tests/providers/amazon/aws/sensors/test_s3_upload_session_complete.py]: Yeah, probably, but I'll update it anyways for future-proofing.

**Comment by anikarni** [comment]
Thanks, @ashb . Good catches, I've made the changes.

As per the naming, I personally like `S3KeysUnchangedSensor` and I agree that the name here is perhaps not the most explicit. However, as I mentioned on the description, this PR was originally a replication of the `GCSUploadSessionCompleteSensor`, so I wasn't sure if we prefer to maintain the naming across providers. Happy to go either way!

**Comment by feluelle** [comment]
> As per the naming, I personally like S3KeysUnchangedSensor and I agree that the name here is perhaps not the most explicit. However, as I mentioned on the description, this PR was originally a replication of the GCSUploadSessionCompleteSensor, so I wasn't sure if we prefer to maintain the naming across providers. Happy to go either way!

I am also in favor of a different name. `S3KeysUnchangedSensor` SGTM :)

File: airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py
Comment (feluelle) [inline airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]:
```suggestion
        self.previous_objects = previous_objects or {}
```

File: airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py
Comment (feluelle) [inline airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]:
```suggestion
    :type previous_objects: Optional[Set[str]]
```
You can use `Optional` in the docs, too.

File: airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py
Comment (feluelle) [inline airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]: Same here.

File: airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py
Comment (feluelle) [inline airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]:
```suggestion
            raise AirflowException(
                'Illegal behavior: objects were deleted in'
                f'{os.path.join(self.bucket, self.prefix)} between pokes.'
            )
```

File: airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py
Comment (feluelle) [inline airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]: Note when you use `"""` it will contain the spaces between lines.

I think it will look like this:
```
SUCCESS:
    Sensor found %s objects at %s.
    Waited at least %s seconds, with no new objects dropped.
```

Comment (feluelle) [reply airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]: So if you want to simply do new lines use `\n` instead.

File: airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py
Comment (feluelle) [reply airflow/providers/amazon/aws/sensors/s3_upload_session_complete.py]: For example here I think the use of `"""` will produce:
```
    Illegal behavior...
```
not sure if it is intended.

The suggestion changes it to:
```
Illegal behavior...
```

**Comment by turbaszek** [comment]
> S3KeysUnchangedSensor SGTM :) @turbaszek WDYT?

 I'm ok with this name

**Comment by anikarni** [comment]
thanks @feluelle ! I've changed the name and implemented your suggestions. If you can give it one more look, I'd certainly appreciate it!

File: tests/providers/amazon/aws/sensors/test_s3_keys_unchanged.py
Comment (feluelle) [inline tests/providers/amazon/aws/sensors/test_s3_keys_unchanged.py]:
```suggestion
        self.sensor.allow_delete = True
```

**Review by feluelle — COMMENTED** [review]
It might makes sense to use [parameterized](https://pypi.org/project/parameterized/) for your tests as some of them only differ in its values - and not really in the testing logic. WDYT?

We have this module already. So you can simply do `from parameterized import parameterized` and use it right away.

**Comment by anikarni** [comment]
@feluelle , I've played around with `parameterized` and pushed the changes here. I myself am a fan of redundancy in tests, as it's useful documentation and the name of the tests makes a whole lot of difference in clarifying intention. But let me know what you think.

Also, not to be too pushy, but this has been quite an investment and I do need to show some progress in this ticket (on our side), so I'm happy to make improvements if they are important, but I also do need to focus on other things. Hope that is understandable ;)

File: tests/providers/amazon/aws/sensors/test_s3_keys_unchanged.py
Comment (feluelle) [inline tests/providers/amazon/aws/sensors/test_s3_keys_unchanged.py]:
```suggestion
        # Test: resetting inactivity period after key change
        (({'a'}, {'a', 'b'}, {'a', 'b', 'c'}), (False, False, False), (0, 0, 0)),
        # ..and in case an item was deleted with option `allow_delete=True`
        (({'a', 'b'}, {'a'}, {'a', 'c'}), (False, False, False), (0, 0, 0)),
        # Test: passes after inactivity period was exceeded
        (({'a'}, {'a'}, {'a'}), (False, False, True), (0, 10, 20)),
        # ..and do not pass if empty key is given
        ((set(), set(), set()), (False, False, False), (0, 10, 20))
```

**Review by feluelle — COMMENTED** [review]
>  I myself am a fan of redundancy in tests, as it's useful documentation and the name of the tests makes a whole lot of difference in clarifying intention. But let me know what you think.

I agree. But in this case I think with a few comments it is pretty readable, what do you think?

> Also, not to be too pushy, but this has been quite an investment and I do need to show some progress in this ticket (on our side), so I'm happy to make improvements if they are important, but I also do need to focus on other things. Hope that is understandable ;)

I totally understand that. But I think you are making progress, don't you? Communicating with eachother is very important in this community. The more people are reviewing a change the more it is "accepted". Also I didn't tell you to change this. It was suggestions I made. I do not insist on them to be accepted. I wanted to know your opinion. That's why I said "WDYT? (what do you think)". However I don't want to be mean.

Let me know what you think of the comments I suggested.

By the way, we need to have @ashb take a look since he did _requested changes_ before he doesn't approve I am not able to merge this.

**Comment by anikarni** [comment]
Yes, of course, just letting you know of my constraints.

> Also please note: I'm going to be away for six weeks at the end of today, so I wont' be able to re-review this. Committers: feel free to dismiss my review once the points have been addressed to your satisfaction.

Sorry, I understood the lack of approval as expecting the changes to be implemented. He had left a message to the effect of his request for changes, so I assumed otherwise ;) thanks for the reviews, anyways!

**Review by feluelle — APPROVED** [review]

**Pull request merged**
